File: frontend/src/physics/constraints.ts
import type { ParticleSystem } from "./ParticleSystem";

export class DistanceConstraintSystem {
  indicesA: Uint16Array;
  indicesB: Uint16Array;
  restLengths: Float32Array;
  stiffnesses: Float32Array;
  count: number;

  constructor() {
    this.indicesA = new Uint16Array(0);
    this.indicesB = new Uint16Array(0);
    this.restLengths = new Float32Array(0);
    this.stiffnesses = new Float32Array(0);
    this.count = 0;
  }

  add(a: number, b: number, restLength: number, stiffness: number = 1.0): void {
    const i = this.count;
    this.count++;

    const grow = <T extends Uint16Array | Float32Array>(
      arr: T,
      val: number,
    ): T => {
      const Constructor = arr.constructor as { new (len: number): T };
      const next = new Constructor(this.count);
      next.set(arr);
      next[i] = val;
      return next;
    };

    this.indicesA = grow(this.indicesA, a);
    this.indicesB = grow(this.indicesB, b);
    this.restLengths = grow(this.restLengths, restLength);
    this.stiffnesses = grow(this.stiffnesses, stiffness);
  }
}

export function solveDistanceConstraints(
  ps: ParticleSystem,
  constraints: DistanceConstraintSystem,
): void {
  if (constraints.count === 0) return;
  const dim = ps.dim;
  const pos = ps.positions;

  // Gauss-Seidel: process one constraint at a time, immediately update positions
  for (let c = 0; c < constraints.count; c++) {
    const a = constraints.indicesA[c];
    const b = constraints.indicesB[c];
    const rest = constraints.restLengths[c];
    const stiffness = constraints.stiffnesses[c];

    let distSq = 0;
    for (let d = 0; d < dim; d++) {
      const dd = pos[b * dim + d] - pos[a * dim + d];
      distSq += dd * dd;
    }
    const dist = Math.sqrt(distSq);
    if (dist < 1e-6) continue;

    const diff = (dist - rest) / dist;
    const halfCorr = stiffness * 0.5 * diff;

    const pinnedA = ps.pinned[a];
    const pinnedB = ps.pinned[b];

    for (let d = 0; d < dim; d++) {
      const delta = pos[b * dim + d] - pos[a * dim + d];
      const correction = halfCorr * delta;
      if (!pinnedA) pos[a * dim + d] += correction;
      if (!pinnedB) pos[b * dim + d] -= correction;
    }
  }
}

export class AngleConstraintSystem {
  indicesA: Uint16Array;
  indicesB: Uint16Array; // pivot
  indicesC: Uint16Array;
  targetAngles: Float32Array;
  stiffnesses: Float32Array;
  count: number;

  constructor() {
    this.indicesA = new Uint16Array(0);
    this.indicesB = new Uint16Array(0);
    this.indicesC = new Uint16Array(0);
    this.targetAngles = new Float32Array(0);
    this.stiffnesses = new Float32Array(0);
    this.count = 0;
  }

  add(a: number, b: number, c: number, targetAngle: number, stiffness: number = 0.5): void {
    const i = this.count;
    this.count++;

    const grow = <T extends Uint16Array | Float32Array>(
      arr: T,
      val: number,
    ): T => {
      const Constructor = arr.constructor as { new (len: number): T };
      const next = new Constructor(this.count);
      next.set(arr);
      next[i] = val;
      return next;
    };

    this.indicesA = grow(this.indicesA, a);
    this.indicesB = grow(this.indicesB, b);
    this.indicesC = grow(this.indicesC, c);
    this.targetAngles = grow(this.targetAngles, targetAngle);
    this.stiffnesses = grow(this.stiffnesses, stiffness);
  }
}

export function solveAngleConstraints(
  ps: ParticleSystem,
  constraints: AngleConstraintSystem,
): void {
  if (constraints.count === 0) return;
  const pos = ps.positions;
  const dim = ps.dim;

  for (let c = 0; c < constraints.count; c++) {
    const a = constraints.indicesA[c];
    const b = constraints.indicesB[c]; // pivot
    const ci = constraints.indicesC[c];
    const target = constraints.targetAngles[c];
    const stiffness = constraints.stiffnesses[c];

    // Vectors from pivot b
    const abx = pos[a * dim] - pos[b * dim];
    const aby = pos[a * dim + 1] - pos[b * dim + 1];
    const cbx = pos[ci * dim] - pos[b * dim];
    const cby = pos[ci * dim + 1] - pos[b * dim + 1];

    const angleAB = Math.atan2(aby, abx);
    const angleCB = Math.atan2(cby, cbx);
    const currentAngle = angleCB - angleAB;

    // Wrap diff to [-pi, pi]
    let diff = currentAngle - target;
    diff = ((diff + Math.PI) % (2 * Math.PI)) - Math.PI;
    // Handle negative modulo
    if (diff < -Math.PI) diff += 2 * Math.PI;

    const correction = diff * stiffness;

    const lenAB = Math.sqrt(abx * abx + aby * aby);
    const lenCB = Math.sqrt(cbx * cbx + cby * cby);
    if (lenAB < 1e-6 || lenCB < 1e-6) continue;

    // Rotate endpoints around pivot
    const newAngleA = angleAB - correction * 0.5;
    const newAngleC = angleCB + correction * 0.5;

    const newAx = pos[b * dim] + Math.cos(newAngleA) * lenAB;
    const newAy = pos[b * dim + 1] + Math.sin(newAngleA) * lenAB;
    const newCx = pos[b * dim] + Math.cos(newAngleC) * lenCB;
    const newCy = pos[b * dim + 1] + Math.sin(newAngleC) * lenCB;

    if (!ps.pinned[a]) {
      pos[a * dim] = newAx;
      pos[a * dim + 1] = newAy;
    }
    if (!ps.pinned[ci]) {
      pos[ci * dim] = newCx;
      pos[ci * dim + 1] = newCy;
    }
  }
}

File: backend/physics/constraints.py
from __future__ import annotations
from dataclasses import dataclass
import torch
from .particle import ParticleSystem


@dataclass
class DistanceConstraintSystem:
    """Hard distance constraints solved via position-based dynamics.
    Ported from traer.py DistanceConstraint.solve (lines 42-54)."""

    indices_a: torch.Tensor     # (C,) int64
    indices_b: torch.Tensor     # (C,) int64
    rest_lengths: torch.Tensor  # (C,) float32
    stiffnesses: torch.Tensor   # (C,) float32

    @staticmethod
    def create() -> DistanceConstraintSystem:
        return DistanceConstraintSystem(
            indices_a=torch.zeros(0, dtype=torch.long),
            indices_b=torch.zeros(0, dtype=torch.long),
            rest_lengths=torch.zeros(0),
            stiffnesses=torch.zeros(0),
        )

    @property
    def count(self) -> int:
        return self.indices_a.shape[0]

    def add(self, a: int, b: int, rest_length: float, stiffness: float = 1.0) -> None:
        self.indices_a = torch.cat([self.indices_a, torch.tensor([a], dtype=torch.long)])
        self.indices_b = torch.cat([self.indices_b, torch.tensor([b], dtype=torch.long)])
        self.rest_lengths = torch.cat([self.rest_lengths, torch.tensor([rest_length])])
        self.stiffnesses = torch.cat([self.stiffnesses, torch.tensor([stiffness])])


def solve_distance_constraints(ps: ParticleSystem, constraints: DistanceConstraintSystem) -> None:
    """Position-based distance constraint solving. Same math as springs but typically
    stiffness=1.0 and iterated multiple times for convergence."""
    if constraints.count == 0:
        return

    pos_a = ps.positions[constraints.indices_a]
    pos_b = ps.positions[constraints.indices_b]
    delta = pos_b - pos_a
    dist = delta.norm(dim=-1, keepdim=True).clamp(min=1e-6)
    diff = (dist - constraints.rest_lengths.unsqueeze(-1)) / dist
    correction = constraints.stiffnesses.unsqueeze(-1) * 0.5 * diff * delta

    pinned_a = ps.pinned[constraints.indices_a].unsqueeze(-1)
    pinned_b = ps.pinned[constraints.indices_b].unsqueeze(-1)
    corr_a = torch.where(pinned_a, torch.zeros_like(correction), correction)
    corr_b = torch.where(pinned_b, torch.zeros_like(correction), correction)

    ps.positions.index_add_(0, constraints.indices_a, corr_a)
    ps.positions.index_add_(0, constraints.indices_b, -corr_b)


@dataclass
class AngleConstraintSystem:
    """Three-particle angle constraints. Particle b is the pivot.
    Ported from traerphysics.js AngleConstraint.apply (lines 176-188)."""

    indices_a: torch.Tensor       # (C,) int64
    indices_b: torch.Tensor       # (C,) int64 — pivot
    indices_c: torch.Tensor       # (C,) int64
    target_angles: torch.Tensor   # (C,) float32
    stiffnesses: torch.Tensor     # (C,) float32

    @staticmethod
    def create() -> AngleConstraintSystem:
        return AngleConstraintSystem(
            indices_a=torch.zeros(0, dtype=torch.long),
            indices_b=torch.zeros(0, dtype=torch.long),
            indices_c=torch.zeros(0, dtype=torch.long),
            target_angles=torch.zeros(0),
            stiffnesses=torch.zeros(0),
        )

    @property
    def count(self) -> int:
        return self.indices_a.shape[0]

    def add(self, a: int, b: int, c: int, target_angle: float, stiffness: float = 0.5) -> None:
        self.indices_a = torch.cat([self.indices_a, torch.tensor([a], dtype=torch.long)])
        self.indices_b = torch.cat([self.indices_b, torch.tensor([b], dtype=torch.long)])
        self.indices_c = torch.cat([self.indices_c, torch.tensor([c], dtype=torch.long)])
        self.target_angles = torch.cat([self.target_angles, torch.tensor([target_angle])])
        self.stiffnesses = torch.cat([self.stiffnesses, torch.tensor([stiffness])])


def solve_angle_constraints(ps: ParticleSystem, constraints: AngleConstraintSystem) -> None:
    """Solve angle constraints by rotating endpoints a and c around pivot b."""
    if constraints.count == 0:
        return

    pos_a = ps.positions[constraints.indices_a]
    pos_b = ps.positions[constraints.indices_b]
    pos_c = ps.positions[constraints.indices_c]

    ab = pos_a - pos_b  # (C, dim)
    cb = pos_c - pos_b  # (C, dim)

    # Compute current angle at pivot b (2D: use atan2)
    angle_ab = torch.atan2(ab[:, 1], ab[:, 0])
    angle_cb = torch.atan2(cb[:, 1], cb[:, 0])
    current_angle = angle_cb - angle_ab

    # Wrap difference to [-pi, pi]
    diff = current_angle - constraints.target_angles
    diff = (diff + torch.pi) % (2 * torch.pi) - torch.pi

    correction = diff * constraints.stiffnesses  # (C,)

    # Rotate a and c by correction angle
    cos_c = torch.cos(correction)
    sin_c = torch.sin(correction)

    # Apply rotation-based positional correction
    len_ab = ab.norm(dim=-1).clamp(min=1e-6)
    len_cb = cb.norm(dim=-1).clamp(min=1e-6)

    # New directions after rotation
    new_angle_a = angle_ab - correction * 0.5
    new_angle_c = angle_cb + correction * 0.5

    new_ab = torch.stack([torch.cos(new_angle_a) * len_ab, torch.sin(new_angle_a) * len_ab], dim=-1)
    new_cb = torch.stack([torch.cos(new_angle_c) * len_cb, torch.sin(new_angle_c) * len_cb], dim=-1)

    new_pos_a = pos_b + new_ab
    new_pos_c = pos_b + new_cb

    # Compute deltas and scatter
    delta_a = new_pos_a - pos_a
    delta_c = new_pos_c - pos_c

    pinned_a = ps.pinned[constraints.indices_a].unsqueeze(-1)
    pinned_c = ps.pinned[constraints.indices_c].unsqueeze(-1)

    delta_a = torch.where(pinned_a, torch.zeros_like(delta_a), delta_a)
    delta_c = torch.where(pinned_c, torch.zeros_like(delta_c), delta_c)

    ps.positions.index_add_(0, constraints.indices_a, delta_a)
    ps.positions.index_add_(0, constraints.indices_c, delta_c)
